### PR TITLE
Fix plantlike mesh size to be horizontally filling a node.

### LINKED ISSUE #1589: New plantlike drawtype causes bad-looking symmetry
Place a plantlike node. Look at it from all 4 directions. 2 of these directions will look poor because of symmetry.

### PR DESCRIPTION
**Pull request by sofar**
fixes #1589: plantlike meshes are only 70% of a node size.

Subsequently, all plantlike textures are stretched by 40% horizontally. Without texture adjustment, this is actually still acceptable, but I intend to submit new textures to minetest_game to repaint all the plantlike textures to the new size. Some will just have extra empty pixels added on the side, but some will get some artwork done to make them look better while retaining the current style.

### CODE REVIEW

**Comment by est31** [comment]
I don't think this should be done. You can already now set `visual_scale` in the nodedef (the jungle grass even does this). So changing this would break behaviour, and leave mod developers helpless because they can't really ensure constant sizes anymore (plantlike is used for lotsa crazy stuff these days, like table legs).

**Comment by sofar** [comment]
As said in the discussion, changing the visual scale to be 1.4x would cause ALL existing plantlike blocks to be 1.4x tall, and extend *above* the 1.0 node height.

This is the right solution. Maybe it's too early for it now, but it really should go in a major release. Visual scale is a far worse solution.

table legs would visually still be entirely within the node block of 1x1x1 with this change, just closer to the edges. With visual_scale at 1.4, the legs would stick out at the top of the table and through a pressure plate. Not what anyone wants. That would *force* everyone to update textures, while with this change it's reasonably similar enough that I bet that a large percentage of players would never notice the difference.

**Comment by 0-afflatus** [comment]
:-1: This should all be possible in Lua. Not convinced.

**Comment by sofar** [comment]
@0-afflatus @est31 I think you're overseeing the likelyhood that the current mesh size is entirely an error - a logical one to make, sure, but still an error. Nodes are supposed to fill a node to it's extent, and not appearing to "fill" only 50% of it's volume!

Fixing the root cause would allow texture painters to further refine the textures if they so design, without having to resort to Lua tricks to get the right scaling. You don't want people to start making 22x22 textures, scale the entire mesh 1.4x and leave the top 6 pixels in each texture empty, that's 3x the work. Whereas with my patch they could choose to leave the textures 16x16, or repaint 22x16, either with transparent pixels on the side to retain the exact same appearance as now, or to move those table legs actually to the corner of the table by using the available extra texture size.

TL:DR: this patch allows more options with less work for everyone.

**Comment by 0-afflatus** [comment]
No, I just don't think it's an error. I'd call that a design decision. Having to repaint textures to get the same effect would be a breakage in my book.

> Nodes are supposed to fill a node to it's extent, and not appearing to "fill" only 50% of it's volume!

That is an entirely false assumption. Don't get me wrong, I like the idea of what you're trying to do, but you must do it so it doesn't break most existing texture packs.

**Comment by paramat** [comment]
:-1: If all plantlike textures are stretched 40% by this commit, how can a 16x16 texture be repainted to make the pixels square again? surely this would require a much-higher resolution texture to enable the 40% horizontal squashing of the texture 'pixels'. It's also a huge amount of work for all textures in all texture packs to be remade.

This is to enable grass to reach a node's corners, which itself is not much needed, the gaps are not a problem and no-one has complained about this before. Individual clumps of grass look better than grass textures that connect to form a grid of grass-walls.

**Comment by 0-afflatus** [comment]
The only way you could do this fairlt AFAICS would be to introduce a new nodetype - "plantlike2" or "grasslike". Most of my plantlike textures were designed to work within the limitations of the plantlike nodetype.

A field

Some table items

I'd need to know these would look better without me having to edit them.

**Comment by paramat** [comment]
For this effect i feel it's better to use the existing per-node 'visual scale' feature and use transparent pixels along the top and bottom of the texture, then existing textures and texture packs are unaffected and pixels remain square. Stretching pixels of all plantlike textures in all texture packs horizontally by 40% would look wrong and be unpopular with texture creators who are being asked to remake their textures with non-square pixels.

**Comment by sofar** [comment]
Good discussion everyone.

I do believe that the introduction of this drawtype was incorrect to begin with, but fixing it does indeed cause many issues for many plantlike nodes.

So I'm closing this - if this was fixed shortly after plantlike drawtype was added then it would have been worth it, but now it's simply too late.

Instead, I'll likely test the waters to see if people are open to scaling the grass nodes up, that'll be a patch to minetest_game.
